### DISCUSSION TITLE
NFT Lesson 17: FlipMoodNFT Interactions error

### DISCUSSION THREAD

**Discussion by mikechaban** [Cyfrin/foundry-full-course-cu#2933] [Q&A] [answered]
Hello there

I am trying to get an additional interaction that is not covered in the lessons to work. After successfully running the MintMoodNFT script on a forked Sepolia chain, I can't run the FlipMoodNFT one. It doesn't even seem like I get a definite error. Here is the error from my terminal

![image](https://github.com/user-attachments/assets/57f37191-388f-49bc-be3e-5ebea7b2dc73)

Then there are some traces and the revert error

![image](https://github.com/user-attachments/assets/29c6592f-9b4c-4a3d-9b5a-10ba5b20d556)

I tried cleaning the project with `forge clean` and deleting the cache directory. I want this script to be a programmatic way to interact with the contract.

```solidity
contract FlipMoodNFT is Script {
    uint256 public constant TOKEN_ID_TO_FLIP = 0;

    function run() external {
        address mostRecentlyDeployedMoodNft = DevOpsTools.get_most_recent_deployment("MoodNFT", block.chainid);
        flipMoodOnContract(mostRecentlyDeployedMoodNft);
    }

    function flipMoodOnContract(address moodNftAddress) public {
        vm.startBroadcast();
        MoodNFT(moodNftAddress).flipMood(TOKEN_ID_TO_FLIP);
        vm.stopBroadcast();
    }
}

```

### [And the script file here](https://github.com/mikechaban/Cyfrin-Updraft/blob/8bb448467e5de157637d9102296b9135ba4f835f/Advanced%20Foundry/nft/script/Interactions.s.sol)

If someone knows how to fix this please let me know

Thank you in advance ❤️

**Comment by EngrPips** [comment]
Hello @mikechaban, What happens if you run an actual sepolia chain and not a forked chain?

**Reply by mikechaban** [reply]
Hi @EngrPips 

I just tried running Sepolia and I get the same error

![image](https://github.com/user-attachments/assets/e68bdaa5-e8a1-40a9-bedd-e6347206ee88)

But then when I run FlipMoodNFT

![image](https://github.com/user-attachments/assets/1825432e-80f5-4743-b684-61f617ccd164)

Do you see any places I could do `console.log` debugging in?

Copilot says that it most probably is either of these:
_1. Verify tokenId Existence
Ensure that the tokenId you are trying to flip exists. If you are using tokenId = 0, make sure that an NFT with tokenId = 0 has been minted.
2. Verify Ownership or Approval
Ensure that the caller of the flipMood function is the owner of the token or has been approved to flip the mood.
3. Check Other Conditions
Ensure that all other conditions in the flipMood function are met._

**Reply by EngrPips** [reply]
Interesting. Does your NFT contract start with a `tokeId` of `0` or `1`? Also, the `0x3FD7...` is the correct address of a MoodNFT instance?

**Reply by mikechaban** [reply]
I assume [my contract](https://github.com/mikechaban/Cyfrin-Updraft/blob/8bb448467e5de157637d9102296b9135ba4f835f/Advanced%20Foundry/nft/src/MoodNFT.sol) starts with `tokenId` of 0 cause I haven't specified it anywhere. I seem to have confirmed this by importing my NFT into Metamask.

![image](https://github.com/user-attachments/assets/6305fcef-7631-4b71-8066-20e07de920bb)

I thought the script creates a new contract every time. Turns out it's been reusing the same `0x3FD7...` contract for several times. You can view it on Etherscan

https://sepolia.etherscan.io/address/0x3FD738Fc27ee9B05F94dc36D8aA1379E3cBBb951

After your response, I checked and noticed there is no transaction other than the contract creation. It means the NFT was not actually minted. The command I was running didn't contain my private key so I included it and the mint transaction appeared on Etherscan.

However, I'm still struggling with the FlipMood interaction. The error I get in the console refers to the `flipMood` function from MoodNFT contract. More specifically, the `flipMood(0)`.
![image](https://github.com/user-attachments/assets/4d2f46c7-bfb4-48de-9d1a-50f15d9cc70b)

I believe the problem is that even though the address IS the owner of the NFT, the address is not "approved", so I'm facing an issue in this part
![image](https://github.com/user-attachments/assets/f154b373-4bd0-4563-a35c-f5a1f0133ce8)

**Reply by mikechaban** [reply]
@EngrPips
I can't say I found the solution, but I found a way to make it work without the approval. I simply got rid of the part highlighted above, created a new contract instance, minted the NFT, and then flipped the mood.

https://sepolia.etherscan.io/address/0x02f0ebac448facc737529f41a1b22cc7f959144e

https://testnets.opensea.io/assets/sepolia/0x02F0ebAc448FacC737529f41a1b22cc7f959144e/0

Does getting rid of this part compromise the security of the contract in terms of confirming exclusively the owner as the address that's able to flip the mood? Based on the code, I can see that only the owner of the token is the one who can flip the mood, so why do we need the first _getApproved_ part from OpenZeppelin's ERC721?

![image](https://github.com/user-attachments/assets/66cb0a96-9a74-406f-ade2-a4c7e8fa841f)

**Reply by EngrPips** [reply]
Please show your `FlipMood` script maybe seeing that can help us make sense of this situation.

**Reply by EngrPips** [reply, marked as answer]
The token can be flipped by the owner or by any other address the owner has approved. Why not just call the `FlipMood` function with the private key that minted the token instead of removing that check?

**Reply by LordAdhaar** [reply]
![image](https://github.com/user-attachments/assets/43a72381-e0de-44d7-947a-79389a01d7e3)
Your transaction might be failing on Sepolia because you haven't properly specified the account or private key. The contract likely can't identify the msg.sender, which is preventing the mood from being flipped. Without clear sender authentication, the transaction won't execute as expected.

Also use `--broadcast` when you do sign the transaction 
You can use environment variables or add `--interactives 1` to the command and it will prompt you to enter the private key and just copy paste that from metamask

**Reply by EngrPips** [reply]
Interesting.

**Reply by mikechaban** [reply]
> The token can be flipped by the owner or by any other address the owner has approved. Why not just call the `FlipMood` function with the private key that minted the token instead of removing that check?

That makes sense now, I didn't realize that the line is there to ensure an approved person can flip as well.

Looks like I forgot to specify the private key and the `--broadcast` flag when running commands. Thanks a lot for the help.

**Reply by EngrPips** [reply]
Awesome, keep crushing it.

**Reply by mikechaban** [reply]
> ![image](https://private-user-images.githubusercontent.com/112782045/390568995-43a72381-e0de-44d7-947a-79389a01d7e3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMxNjk3NzAsIm5iZiI6MTczMzE2OTQ3MCwicGF0aCI6Ii8xMTI3ODIwNDUvMzkwNTY4OTk1LTQzYTcyMzgxLWUwZGUtNDRkNy05NDdhLTc5Mzg5YTAxZDdlMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwMlQxOTU3NTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iOGQ3MzM4OTFjYmEwZWY0NGEyOTk5NjlkNGYzYmU4MjJkOTM2NGE2ZDU5MzNiOGJlY2VjNzZlM2JjZDMyNDVkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.UiiaRbT_5VdufyYT676vbD93ohp_1z3I0BgMS2IGmeM) Your transaction might be failing on Sepolia because you haven't properly specified the account or private key. The contract likely can't identify the msg.sender, which is preventing the mood from being flipped. Without clear sender authentication, the transaction won't execute as expected.
> 
> Also use `--broadcast` when you do sign the transaction You can use environment variables or add `--interactives 1` to the command and it will prompt you to enter the private key and just copy paste that from metamask

You're exactly right - I forgot to specify the private key and the `--broadcast` flag when running commands. Thank you for the detailed explanation.

**Reply by LordAdhaar** [reply]
Interesting. Happy to help!
